### ISSUE TITLE
rowRenderer onclick

### ISSUE THREAD

**Issue by denkomanceski** [cheton/infinite-tree#9]
When I click on a node, the row renderer is called twice, with two different NodeIds. It is called with the previous selected node, and the one that should be selected.
![gif](https://cloud.githubusercontent.com/assets/10501247/23793980/2d115446-058e-11e7-8411-22ebc62f0ed3.gif)

And sometimes, when I updateNode, I change the id of the current node, and in my app its trying to call the function with the new id (which is good), and than with the old id (which doesnt exist anymore, and the app is throwing error).

As you can see from the gif, the click event is fired only once (thats how it should be), and the console.log inside of the rowRenderer is called twice.

So if I call "update" on a node, is it also trying to update all of its children ?

**Comment by cheton** [comment]
A click event will call row renderer twice, this is the expected behavior. The first call is to render the next selected node, the second call is to render previous selected node so it will display the unselected state:

```js
// Select this node
if (this.state.selectedNode !== node) {
    node.state.selected = true;

    // Update the row corresponding to the node
    this.rows[nodeIndex] = this.options.rowRenderer(node, this.options);
}

// Deselect the current node
if (this.state.selectedNode) {
    const selectedNode = this.state.selectedNode;
    const selectedIndex = this.nodes.indexOf(this.state.selectedNode);
    selectedNode.state.selected = false;
    this.rows[selectedIndex] = this.options.rowRenderer(selectedNode, this.options);
}
```

In your case, the updateNode method does not remove unused id from the lookup table (e.g. `this.nodeTable`) because I didn't expect one would ever change the node id. I think that would be the problem that caused an error.

Lets take a look at getNodeById, it will lookup the node by id from the nodeTable, if an unused node id was not removed from the nodeTable, it will return the node immediately, and therefore caused unexpected behavior.

```js
 getNodeById(id) {
    let node = this.nodeTable.get(id);
    if (!node) {
        // Find the first node that matches the id
        node = this.nodes.filter((node) => {
            return (node.id === id);
        })[0];
        if (!node) {
            return null;
        }
        this.nodeTable.set(node.id, node);
    }
    return node;
}
```

**Comment by cheton** [comment]
The lookup table should be refreshed when changing the node id with updateNode. I will fix it in the 1.9 release.

**Comment by cheton** [comment]
Fixed in v1.8.2
